File: frontend/src/services/review.ts
import { apiRequest } from "./api";
import type { Frame } from "../types/frame";
import type { Label, ReviewSession } from "../types/review";

export const listLabels = async (projectId: number, signal?: AbortSignal) =>
  (
    await apiRequest<{ data: Label[] }>(`/projects/${projectId}/labels`, {
      signal,
    })
  ).data;
export const createLabel = async (
  projectId: number,
  input: { name: string; shortcut: string | null; color: string },
) =>
  (
    await apiRequest<{ data: Label }>(`/projects/${projectId}/labels`, {
      method: "POST",
      body: JSON.stringify(input),
    })
  ).data;
export const updateLabel = async (
  id: number,
  input: Partial<
    Pick<Label, "name" | "shortcut" | "color" | "description" | "position">
  >,
) =>
  (
    await apiRequest<{ data: Label }>(`/labels/${id}`, {
      method: "PATCH",
      body: JSON.stringify(input),
    })
  ).data;
export const deleteLabel = (id: number) =>
  apiRequest<void>(`/labels/${id}`, { method: "DELETE" });
export const assignLabels = async (frameId: number, labelIds: number[]) =>
  (
    await apiRequest<{ data: Frame }>(`/frames/${frameId}/labels`, {
      method: "POST",
      body: JSON.stringify({ label_ids: labelIds }),
    })
  ).data;
export const removeLabel = async (frameId: number, labelId: number) =>
  (
    await apiRequest<{ data: Frame }>(`/frames/${frameId}/labels/${labelId}`, {
      method: "DELETE",
    })
  ).data;
export const bulkLabels = (
  frameIds: number[],
  labelIds: number[],
  action: "assign" | "remove",
) =>
  apiRequest<void>("/frames/bulk-label", {
    method: "POST",
    body: JSON.stringify({ frame_ids: frameIds, label_ids: labelIds, action }),
  });
export type ReviewChanges = {
  review_status?: string;
  favorite?: boolean;
  rejected?: boolean;
};
export const reviewFrame = async (frameId: number, changes: ReviewChanges) =>
  (
    await apiRequest<{ data: Frame }>(`/frames/${frameId}/review`, {
      method: "PATCH",
      body: JSON.stringify(changes),
    })
  ).data;
export const bulkReview = (frameIds: number[], changes: ReviewChanges) =>
  apiRequest<void>("/frames/bulk-review", {
    method: "POST",
    body: JSON.stringify({ frame_ids: frameIds, ...changes }),
  });
export interface BatchTarget {
  frame_ids: number[];
  all_filtered: boolean;
  filters: Record<string, string | string[]>;
}
export const filteredBulkLabels = (
  projectId: number,
  target: BatchTarget,
  labelIds: number[],
  action: "assign" | "remove",
) =>
  apiRequest<{ affected_count: number }>(
    `/projects/${projectId}/frames/bulk-label`,
    {
      method: "POST",
      body: JSON.stringify({
        ...target,
        label_ids: labelIds,
        action,
      }),
    },
  );
export const filteredBulkReview = (
  projectId: number,
  target: BatchTarget,
  changes: ReviewChanges,
) =>
  apiRequest<{ affected_count: number }>(
    `/projects/${projectId}/frames/bulk-review`,
    {
      method: "POST",
      body: JSON.stringify({ ...target, ...changes }),
    },
  );
export const getSession = async (projectId: number, signal?: AbortSignal) =>
  (
    await apiRequest<{ data: ReviewSession }>(
      `/projects/${projectId}/review-session`,
      { signal },
    )
  ).data;
export const saveSession = async (
  projectId: number,
  changes: Record<string, unknown>,
) =>
  (
    await apiRequest<{ data: ReviewSession }>(
      `/projects/${projectId}/review-session`,
      { method: "PATCH", body: JSON.stringify(changes) },
    )
  ).data;

File: backend/app/api/routes/review.py
from typing import Annotated

from fastapi import APIRouter, Depends, Request, Response
from sqlalchemy.orm import Session

from app.database.session import get_db
from app.schemas.frame import FrameEnvelope
from app.schemas.review import (
    BulkActionEnvelope,
    BulkActionResult,
    BulkLabelUpdate,
    BulkReviewUpdate,
    FilteredLabelUpdate,
    FilteredReviewUpdate,
    FrameLabelsUpdate,
    LabelCreate,
    LabelEnvelope,
    LabelListEnvelope,
    LabelRead,
    LabelUpdate,
    ReviewSessionEnvelope,
    ReviewSessionUpdate,
    ReviewUpdate,
)
from app.services.frame import FrameService
from app.services.project import ProjectService
from app.services.review import LabelService, ReviewService, ReviewSessionService

router = APIRouter(tags=["review"])


def db(session: Annotated[Session, Depends(get_db)]) -> Session:
    return session


@router.get("/projects/{project_id}/labels", response_model=LabelListEnvelope)
def labels(project_id: int, session: Annotated[Session, Depends(db)]) -> LabelListEnvelope:
    return LabelListEnvelope(
        data=[LabelRead.model_validate(label) for label in LabelService(session).list(project_id)]
    )


@router.post("/projects/{project_id}/labels", response_model=LabelEnvelope, status_code=201)
def create_label(
    project_id: int,
    payload: LabelCreate,
    request: Request,
    session: Annotated[Session, Depends(db)],
) -> LabelEnvelope:
    ProjectService(session, request.app.state.settings.storage_root).get(project_id)
    return LabelEnvelope(
        data=LabelRead.model_validate(LabelService(session).create(project_id, payload))
    )


@router.patch("/labels/{label_id}", response_model=LabelEnvelope)
def update_label(
    label_id: int, payload: LabelUpdate, session: Annotated[Session, Depends(db)]
) -> LabelEnvelope:
    return LabelEnvelope(
        data=LabelRead.model_validate(LabelService(session).update(label_id, payload))
    )


@router.delete("/labels/{label_id}", status_code=204)
def delete_label(label_id: int, session: Annotated[Session, Depends(db)]) -> Response:
    LabelService(session).delete(label_id)
    return Response(status_code=204)


def frame_envelope(frame_id: int, request: Request, session: Session) -> FrameEnvelope:
    frames = FrameService(session, request.app.state.settings.storage_root)
    return FrameEnvelope(data=frames.serialize(frames.get(frame_id)))


@router.post("/frames/{frame_id}/labels", response_model=FrameEnvelope)
def assign_labels(
    frame_id: int,
    payload: FrameLabelsUpdate,
    request: Request,
    session: Annotated[Session, Depends(db)],
) -> FrameEnvelope:
    ReviewService(session).labels(frame_id, payload.label_ids)
    return frame_envelope(frame_id, request, session)


@router.delete("/frames/{frame_id}/labels/{label_id}", response_model=FrameEnvelope)
def remove_label(
    frame_id: int, label_id: int, request: Request, session: Annotated[Session, Depends(db)]
) -> FrameEnvelope:
    ReviewService(session).remove_label(frame_id, label_id)
    return frame_envelope(frame_id, request, session)


@router.post("/frames/bulk-label", status_code=204)
def bulk_label(payload: BulkLabelUpdate, session: Annotated[Session, Depends(db)]) -> Response:
    ReviewService(session).bulk_labels(payload.frame_ids, payload.label_ids, payload.action)
    return Response(status_code=204)


@router.patch("/frames/{frame_id}/review", response_model=FrameEnvelope)
def review_frame(
    frame_id: int, payload: ReviewUpdate, request: Request, session: Annotated[Session, Depends(db)]
) -> FrameEnvelope:
    ReviewService(session).review(frame_id, payload)
    return frame_envelope(frame_id, request, session)


@router.post("/frames/bulk-review", status_code=204)
def bulk_review(payload: BulkReviewUpdate, session: Annotated[Session, Depends(db)]) -> Response:
    ReviewService(session).bulk_review(
        payload.frame_ids, ReviewUpdate(**payload.model_dump(exclude={"frame_ids"}))
    )
    return Response(status_code=204)


@router.post("/projects/{project_id}/frames/bulk-label", response_model=BulkActionEnvelope)
def filtered_bulk_label(
    project_id: int,
    payload: FilteredLabelUpdate,
    request: Request,
    session: Annotated[Session, Depends(db)],
) -> BulkActionEnvelope:
    service = ReviewService(session)
    ids = service.resolve_target(
        project_id,
        payload.frame_ids,
        payload.all_filtered,
        payload.filters,
        request.app.state.settings.storage_root,
    )
    service.bulk_labels(ids, payload.label_ids, payload.action)
    return BulkActionEnvelope(data=BulkActionResult(affected_count=len(ids)))


@router.post("/projects/{project_id}/frames/bulk-review", response_model=BulkActionEnvelope)
def filtered_bulk_review(
    project_id: int,
    payload: FilteredReviewUpdate,
    request: Request,
    session: Annotated[Session, Depends(db)],
) -> BulkActionEnvelope:
    service = ReviewService(session)
    ids = service.resolve_target(
        project_id,
        payload.frame_ids,
        payload.all_filtered,
        payload.filters,
        request.app.state.settings.storage_root,
    )
    service.bulk_review(
        ids,
        ReviewUpdate(**payload.model_dump(exclude={"frame_ids", "all_filtered", "filters"})),
    )
    return BulkActionEnvelope(data=BulkActionResult(affected_count=len(ids)))


@router.get("/projects/{project_id}/review-session", response_model=ReviewSessionEnvelope)
def get_review_session(
    project_id: int, session: Annotated[Session, Depends(db)]
) -> ReviewSessionEnvelope:
    service = ReviewSessionService(session)
    return ReviewSessionEnvelope(data=service.serialize(service.get(project_id)))


@router.patch("/projects/{project_id}/review-session", response_model=ReviewSessionEnvelope)
def update_review_session(
    project_id: int, payload: ReviewSessionUpdate, session: Annotated[Session, Depends(db)]
) -> ReviewSessionEnvelope:
    service = ReviewSessionService(session)
    return ReviewSessionEnvelope(data=service.serialize(service.update(project_id, payload)))
